[MRG] Issue 9698 safe and short repr

lebigot opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Contains the basis for fixing #9698.

#### What does this implement/fix? Explain your changes.

Utility functions that provide safe and friendlier versions of str() and repr() were added to `utils`: `safe_str()` and `safe_repr()`.

These functions _should never crash_ and should _always return a string_.

They are intended to be used in `raise` statements and warning calls. I created a **list that should contain most (if not all) such places**: https://gist.github.com/lebigot/8526af2425355b5cf837068c4182fc2d.

I have thus started to go through this list (in the listed order): this pull requests contains the first changes.

#### Any other comments?

I considered that there is no need to protect the representation of something like `arr.shape` when `arr` is supposed to be a NumPy array or equivalent (and has likely be used as such before the error message).

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

lebigot committed:
Added utils.safe_str() and utils.nice_repr().

safe_str() is a safe version of str(): it always returns a string and
doesn't crash (even if str() would crash).

nice_repr() is similar but for repr(), and in addition provides a nicer
representation if possible (with a limited width, etc.).

lebigot committed:
nice_repr() now limits the size of the returned text.

lebigot committed:
Renamed nice_repr() -> safe_repr().

This follows the sklearn convention of name "safe_*()" functions
that do a bit more than being safe (Gaël Varoquaux).

lebigot committed:
First application of the safe_repr() representation.

This is the first function found at
https://gist.github.com/lebigot/8526af2425355b5cf837068c4182fc2d.

lebigot committed:
Using utils.safe_repr() in more funtions.

Still following the list at
https://gist.github.com/lebigot/8526af2425355b5cf837068c4182fc2d.

lebigot committed:
Protection via utils.safe_repr() added.

Still going through the list at
https://gist.github.com/lebigot/8526af2425355b5cf837068c4182fc2d.

lebigot committed:
Used safe_repr() for taking precautions in error message.

I am still following the list of potential places for safe_repr()
in exceptions and warnings at
https://gist.github.com/lebigot/8526af2425355b5cf837068c4182fc2d.

lebigot committed:
Added protection of function argument through safe_repr().

Still going through the list at
https://gist.github.com/lebigot/8526af2425355b5cf837068c4182fc2d.

lebigot committed:
More safe_repr() for error messages.

Other %r found in the modified file should often work because they are applied
to quantities that were just successfully compared to an integer.

lebigot committed:
safe_repr() protection added.

Still going through
https://gist.github.com/lebigot/8526af2425355b5cf837068c4182fc2d.

GaelVaroquaux commented on sklearn/utils/__init__.py:
```diff
@@ -590,3 +591,84 @@ def is_scalar_nan(x):
     # Redondant np.floating is needed because numbers can't match np.float32
     # in python 2.
     return bool(isinstance(x, (numbers.Real, np.floating)) and np.isnan(x))
+
```
PEP8 says that you should have 2 line returns here

lebigot replied:
Addressed in https://github.com/lebigot/scikit-learn/commit/6bc2a39fe5e6a9f65c9d54fb442e18a265e82e9a

GaelVaroquaux commented on sklearn/utils/__init__.py:
```diff
@@ -590,3 +591,84 @@ def is_scalar_nan(x):
+        return '{} instance'.format(safe_str(klass))
+    else:
+        raise AttributeError("No __name__ or __class__ attribute.")
+
```
Here too: 2 line returns

lebigot replied:
Addressed in https://github.com/lebigot/scikit-learn/commit/6bc2a39fe5e6a9f65c9d54fb442e18a265e82e9a.

GaelVaroquaux commented on sklearn/utils/__init__.py:
```diff
@@ -590,3 +591,84 @@ def is_scalar_nan(x):
+    If no callable succeeds, returns no_transform.
+
+    transformers -- sequence of 1-parameter callables.
+    no_transform -- returned if no transformer function succeeds.
```
You should use the numpydoc conventions here.

lebigot replied:
Addressed in https://github.com/lebigot/scikit-learn/commit/6bc2a39fe5e6a9f65c9d54fb442e18a265e82e9a.

GaelVaroquaux commented on sklearn/utils/__init__.py:
```diff
@@ -590,3 +591,84 @@ def is_scalar_nan(x):
+    else:
+        raise AttributeError("No __name__ or __class__ attribute.")
+
+def safe_version(transformers, no_transform):
```
I would call transformers "callables" and "no_transform" => "no_operation"

GaelVaroquaux replied:
Also, I would call "safe_version" => "safe_call"

lebigot replied:
`no_transform` is actually the _result_ of the returned function if none of the calls succeeds: it's not a no-op, so I am proposing to call it `default_result` instead.

lebigot replied:
Addressed in https://github.com/lebigot/scikit-learn/commit/6bc2a39fe5e6a9f65c9d54fb442e18a265e82e9a.

GaelVaroquaux commented on sklearn/utils/__init__.py:
```diff
@@ -590,3 +591,84 @@ def is_scalar_nan(x):
+        raise AttributeError("No __name__ or __class__ attribute.")
+
+def safe_version(transformers, no_transform):
+    """Return a function that applies the first working callable.
```
"The first working callable" => "The first callable that does not raise an error".

lebigot replied:
I wanted to do this, but then we go beyond the 79-character limit of PEP 8.

I am thus changing this to "the first callable that returns a value".

The details should be clear in the docstring, in any case.

lebigot replied:
Addressed in https://github.com/lebigot/scikit-learn/commit/6bc2a39fe5e6a9f65c9d54fb442e18a265e82e9a.

GaelVaroquaux commented on sklearn/utils/__init__.py:
```diff
@@ -590,3 +591,84 @@ def is_scalar_nan(x):
+    # documentation, so it may have to be replaced by something else:
+    (pydoc.text.repr, repr, safe_name_or_class), "<No object representation>")
+
+MAX_TEXT_LENGTH = 80*100  # 100 standard lines
```
Empty line here

lebigot replied:
Addressed in https://github.com/lebigot/scikit-learn/commit/6bc2a39fe5e6a9f65c9d54fb442e18a265e82e9a.

GaelVaroquaux reviewed (commented):
Looks great!!!

Are you going to work on adding the function to the rest of code?

lebigot committed:
Added guideline on the use of safe_str() and safe_repr().

lebigot commented:
Yes, I am planning to go through the list of places where the change should be made.

lebigot committed:
Addressed comments by GaelVaroquaux.

https://github.com/scikit-learn/scikit-learn/pull/11601

lebigot committed:
One additional protected function.

lebigot committed:
One additional file with safe representations.

Places where the code should often crash before requiring the representation
of an object were not protected.

lebigot committed:
One additional file with safe representations.

Places where the code should often crash before requiring the representation
of an object were not protected.

lebigot committed:
One additional file with safe representations.

Places where the code should often crash before requiring the representation
of an object were not protected.

lebigot committed:
One additional file with safe representations.

Places where the code should often crash before requiring the representation
of an object were not protected.

lebigot committed:
One additional file with safe representations.

Places where the code should often crash before requiring the representation
of an object were not protected.

lebigot committed:
More files with utils.safe_repr() where needed.

Places where the code should often crash before requiring the representation
of an object were not protected.

lebigot committed:
One additional file with safe representations.

Places where the code should often crash before requiring the representation
of an object were not protected.

lebigot committed:
More files with safe string representations.

Places where the code should often crash before requiring the representation
of an object were not protected.

lebigot committed:
More files with safe representations

Places where the code should often crash before requiring the representation
of an object were not protected. Similarly, places where the object
being represented has little chance of being of an exotic type are not
protected.

lebigot committed:
Final set of utils.safe_repr() wrappings.

Not all representations were wrapped (see previous commit messages),
because sometimes the type of the represented variable is quite
constrained and should not generate any problem with the representation
string (maybe it is explicit, maybe the code already expects that it is
numerical, maybe some specific attributes were just tested, which is
already a pretty good test of the type, etc.).

lebigot committed:
Merge branch 'master' into issue_9698_safe_and_short_repr

lebigot commented:
I went through the list of `raise` that use the Python representation of variables (https://gist.github.com/lebigot/8526af2425355b5cf837068c4182fc2d) and wrapped the relevant objects with `utils.safe_repr()`.

Not all representations were wrapped: sometimes this does not appear to be fully needed because sometimes the type of the represented variable is quite constrained and should not generate any problem with the representation string (maybe it is explicit, maybe the code already expects that it is numerical, maybe some specific attributes were just tested, which is already a pretty good test of the type, etc.). Only some _very_ unlikely cases of crashing representations should remain.

TomDLT commented on sklearn/datasets/samples_generator.py:
```diff
@@ -790,7 +791,7 @@ def make_blobs(n_samples=100, n_features=2, centers=None, cluster_std=1.0,
         except TypeError:
             raise ValueError("Parameter `centers` must be array-like. "
-                             "Got {!r} instead".format(centers))
+                             "Got {!r} instead".format(safe_repr(centers)))
```
Change to {!s} ?

TomDLT commented on sklearn/decomposition/nmf.py:
```diff
@@ -1001,13 +1001,13 @@ def non_negative_factorization(X, W=None, H=None, n_components=None,
     if not isinstance(tol, numbers.Number) or tol < 0:
         raise ValueError("Tolerance for stopping criteria must be "
-                         "positive; got (tol=%r)" % tol)
+                "positive; got (tol=%s)" % safe_repr(tol))
```
PEP 8

TomDLT commented on sklearn/utils/__init__.py:
```diff
@@ -597,3 +598,99 @@ def is_scalar_nan(x):
+safe_str = safe_call((str,), "<No string representation>")
+
+
+# The function nice_repr() should be used when we want a more user-friendly
```
The comment is outdated, there is no `nice_repr` function.

TomDLT commented on sklearn/utils/__init__.py:
```diff
@@ -597,3 +598,99 @@ def is_scalar_nan(x):
+
+
+MAX_TEXT_LENGTH = 80*100  # 100 standard lines
+def max_length_nice_repr(obj):
```
I would always start function names with `safe_repr_` for consistency.

TomDLT commented on sklearn/utils/__init__.py:
```diff
@@ -597,3 +598,99 @@ def is_scalar_nan(x):
+    (pydoc.text.repr, repr, safe_name_or_class), "<No object representation>")
+
+
+MAX_TEXT_LENGTH = 80*100  # 100 standard lines
```
PEP8: `80 * 100`

TomDLT commented on sklearn/utils/tests/test_utils.py:
```diff
@@ -22,6 +22,7 @@
 from sklearn.utils import gen_even_slices
 from sklearn.utils import get_chunk_n_rows
 from sklearn.utils import is_scalar_nan
+from sklearn.utils import safe_str, nice_repr
```
`nice_repr` does not exist.

TomDLT commented on sklearn/utils/__init__.py:
```diff
@@ -597,3 +598,99 @@ def is_scalar_nan(x):
+
+
+# We're taking precautions by wrapping max_length_nice_repr() with
+# safe_version(), but if we trust its code, we don't have to. This would
```
outdated? what is `safe_version`?

TomDLT reviewed (commented):
Thanks for this PR.

You have conflicts, you should merge master into this branch and solve the conflicts.
You also have import errors (see job 1, 2, 3 in [Travis report](https://travis-ci.org/scikit-learn/scikit-learn/builds/415198902?utm_source=github_status&utm_medium=notification))
You also have PEP8 warnings (see [job 4 in Travis](https://travis-ci.org/scikit-learn/scikit-learn/jobs/415198906)).

Also some minor comments.

amueller commented:
how does this interact with #11705?
